contributors/design-proposals/ipvs-proxy.md

```diff
@@ -0,0 +1,152 @@
+
+### Port remapping 
+
+There are 3 proxy mode in ipvs, NAT, IPIP and DR. Only NAT mode support port remapping.
+
+### NodePort type service support
+
+For NodePort type service, IPVS proxy will take all accessable IPs in a host as the virtual IP of ipvs service. Specifically, accessable IP excludes `lo`, `docker0`, `vethxxx`, `cni0`, `flannel0`, etc. Currently, we assume they are IPs bound to `eth{i}`.
+
+We people visit `NodeIP:NodePort`, it will hit the VIP of ipvs service. IPVS proxy doesn't need to assign Node IP to dummy interface.
+
+### LoadBalancer type service support
+
+For LoadBalancer type service, IPVS proxy will take the loadbalancer's ingress IP as the virtual IP of ipvs service. 
+
+Since there is a need of supporting access control for ingress IP. IPVS proxy will make use of iptables to do access control, Like
+
+```shell
+-I KUBE-SERVICES -d {ingress.IP} --dport {service.Port} -s {LoadBalancerSource} -j ACCEPT
+```
+
+When the packet reach the end of firewall chain and does not get DNATed, IPVS proxy will drop it, iptables rule like
+
+```shell
+-A KUBE-SERVICES -d {ingress.IP} --dport {service.Port} -j KubeMarkDropChain
+```
+
+### Support service cluster IP
+
+IPVS proxy will assign service cluster IP to the dummy interface on each cluster node. Specifically,  it will create an alias on dummy interface for each cluster IP. Delete a service will trigger the deletion of alias device.
+
+### Support service external IP
+
+IPVS proxy will not create alias device for external IPs of a service. It only create a ipvs service whose VIP corresponding to external IP.
+
+### Session affinity
+
+ipvs support client IP session affinity, it's called persistent connection. It's the `-p` flag in ipvsadm, like
+
+```shell
+ipvsadm -A -t 10.244.1.100:8080 -s rr -p [timeout]
+```
+
+When a service specify session affinity, ipvs proxy will assign a timeout value(180min by default) to ipvs service.
+
+### Sync period
+
+Similar to iptables proxy, IPVS proxy will do full sync loop every 10 seconds by default. Besides, every update on kubernetes service and endpoint will trigger an ipvs service and destination update.
+
+### Support Only NodeLocal Endpoints
+
+Similar to iptables proxy, IPVS proxy support Only NodeLocal Endpoints feature. When a service has the annotation,  ipvs proxy will do traffic proxy only for local endpoints.
+
+### Network policy
+
+For IPVS NAT mode to work, **all packets from the realservers to the client must go through the director**. It means ipvs proxy should do SNAT in a L3-overlay network(such as flannel) for cross-host network communication. When a container request a cluster IP to visit an endpoint in another host, we should enable `--masquer-all` for ipvs proxy, **which will break network policy**.
```

Comment:
apart from soon to be a GA feature, network policy is really important that people probably would trade performance for it.  A draft idea is to not do snat, rather, on each host, add a new routing table and fwmark all service traffic to this table. The routing table will take care of routing back the packet.  There's a lot of caveats, i don't know if it will ever work at all.